panda-rs/src/api/rr.rs

```diff
@@ -0,0 +1,57 @@
+use std::os::raw::c_int;
+use std::ffi::CString;
+use std::ptr;
+
+use crate::Error;
+
+/// RR point-in-time: get current count of instructions replayed
+pub fn rr_get_guest_instr_count() -> c_int {
+    unsafe {
+        panda_sys::rr_get_guest_instr_count_external()
+    }
+}
+
+
+/// Stop and quit, wraps QMP functions.
+pub fn vm_quit() -> c_int {
+    unsafe {
+        panda_sys::panda_replay_end()
+    }
+}
+
+/// Start recording.
+/// If `snapshot.is_some()` restore the named snapshot prior to recording.
+pub fn record_begin(name: &str, snapshot: Option<&str>) -> Result<c_int, Error> {
```

Comment:
Converting this c_int into a Result itself would probably be preferable, as the int has 3 values: `RRCTRL_EINVALID` (recording already on, `panda_sys::RRCTRL_ret_RRCTRL_EINVALID`), `RRCTRL_EPENDING` (recording in the process of being turned on already, `panda_sys::RRCTRL_ret_RRCTRL_EINVALID`) and `OK` (`panda_sys::RRCTRL_ret_RRCTRL_OK`).

https://github.com/panda-re/panda/blob/ecfe828a302cdd99cf0858063d27624c49dcb3f1/panda/src/callbacks.c#L767-L770

If you don't feel like cramming this in the big error, you could also make an `RrError` enum or something similar.

Reply:
Not propagating up `c_int`s is a really good call, thanks!

Went with an `RrError` `enum` that handles translation from `c_int` via an associated method. The existing top-level `Error` now uses `#[error(transparent)]` to pass `RrError` through. Hope this patterns makes it easy to add other "subclasses" of errors in the future while maintaining the same external API!

All relevant return signatures updated from `Result<c_int, Error>` to `Result<(), Error>` where `Ok` == `panda_sys::RRCTRL_ret_RRCTRL_OK`.